src/ua.c

```diff
@@ -2020,6 +2113,73 @@ struct ua *uag_find(const struct pl *cuser)
 }
 
 
+/**
+ * Find the correct UA from SIP message
+ *
+ * @param msg SIP message
+ *
+ * @return Matching UA if found, NULL if not found
+ */
+struct ua *uag_find_msg(const struct sip_msg *msg)
+{
+	struct le *le;
+	const struct pl *cuser;
+	struct ua *uaf = NULL;  /* fallback ua */
+
+	if (!msg)
+		return NULL;
+
+	cuser = &msg->uri.user;
+	for (le = uag.ual.head; le; le = le->next) {
+		struct ua *ua = le->data;
+
+		if (0 == pl_strcasecmp(cuser, ua->cuser)) {
+			ua_printf(ua, "selected for %r\n", cuser);
+			return ua;
+		}
+	}
+
+	/* Try also matching by AOR, for better interop and for peer-to-peer
+	 * calls */
+	for (le = uag.ual.head; le; le = le->next) {
+		struct ua *ua = le->data;
+		struct account *acc = ua->acc;
+
+		if (!uri_match_transport(&acc->luri, msg->tp))
+			continue;
+
+		if (!uri_match_af(&acc->luri, &msg->uri))
+			continue;
+
+		if (uri_host_local(&acc->luri) != uri_host_local(&msg->uri))
+			continue;
```

Comment:
I am wondering why we need to compare the IP-address of the incoming request?

the SIP stack is listening to an interface:

```
transports:
  10.0.1.13:56669 (UDP)
  10.0.1.13:56726 (TCP)
  10.0.1.13:56727 (TLS)
  10.0.1.13:9 (WS)
  10.0.1.13:9 (WSS)
```

an incoming request to 10.0.1.13 will be handled by the SIP-stack, and then
passed up to the application. incoming IP packets to different address/port will be
dropped by the kernel.

so I am not sure if I understand the usecase, say you have 3 uas:

```
<sip:user@10.0.1.13>
<sip:user@127.0.0.1>
<sip:user@localhost>
```

you want the incoming request to arrive at the correct UA?
is it perhaps enough to just use one UA?